Deploy and manage: Autoscaling migration #344 (#606)
elastic/docs-content: deploy-manage/autoscaling/trained-model-autoscaling.md

## Enabling autoscaling in {{kib}} - adaptive resources [enabling-autoscaling-in-kibana-adaptive-resources]

You can enable adaptive resources for your models when starting or updating the model deployment. Adaptive resources make it possible for {{es}} to scale up or down the available resources based on the load on the process. This can help you to manage performance and cost more easily. When adaptive resources are enabled, the number of vCPUs that the model deployment uses is set automatically based on the current load. When the load is high, the number of vCPUs that the process can use is automatically increased. When the load is low, the number of vCPUs that the process can use is automatically decreased.

You can choose from three levels of resource usage for your trained model deployment; autoscaling will occur within the selected level’s range.

Refer to the tables in the [Model deployment resource matrix](#model-deployment-resource-matrix) section to find out the setings for the level you selected.

:::{image} ../../images/machine-learning-ml-nlp-deployment-id-elser-v2.png
:alt: ELSER deployment with adaptive resources enabled.
:class: screenshot
:width: 500px
:::

> Search projects are given access to more processing resources, while Security and Observability projects have lower limits. This difference is reflected in the UI configuration: Search projects have higher resource limits compared to Security and Observability projects to accommodate their more complex operations.

Reviewer: ```suggestion
In {{serverless-full}}, Search projects are given access to more processing resources, while Security and Observability projects have lower limits. This difference is reflected in the UI configuration: Search projects have higher resource limits compared to Security and Observability projects to accommodate their more complex operations.
```

On {{serverless-short}}, adaptive allocations are automatically enabled for all project types. However, the "Adaptive resources" control is not displayed in Kibana for Observability and Security projects.

## Model deployment resource matrix [model-deployment-resource-matrix]

The used resources for trained model deployments depend on three factors:

> * your cluster environment ({{serverless-short}}, Cloud, or on-premises)
> * the use case you optimize the model deployment for (ingest or search)
> * whether model autoscaling is enabled with adaptive allocations/resources to have dynamic resources, or disabled for static resources
> 
> If you use {{es}} on-premises, vCPUs level ranges are derived from the `total_ml_processors` and `max_single_ml_node_processors` values. Use the [get {{ml}} info API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-ml-info) to check these values. The following tables show you the number of allocations, threads, and vCPUs available in Cloud when adaptive resources are enabled or disabled.

Reviewer: ```suggestion
* your cluster environment ({{serverless-short}}, Cloud (ECE, ECK, ECH), or self-managed)
* the use case you optimize the model deployment for (ingest or search)
* whether model autoscaling is enabled with adaptive allocations/resources to have dynamic resources, or disabled for static resources

If you use a self-managed cluster, vCPUs level ranges are derived from the `total_ml_processors` and `max_single_ml_node_processors` values. Use the [get {{ml}} info API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-ml-info) to check these values. The following tables show you the number of allocations, threads, and vCPUs available in ECE, ECK, and ECH when adaptive resources are enabled or disabled.
```

::::{note}
On {{serverless-short}}, adaptive allocations are automatically enabled for all project types. However, the "Adaptive resources" control is not displayed in {{kib}} for Observability and Security projects.
::::

### Ingest optimized

In case of ingest-optimized deployments, we maximize the number of model allocations.

#### Adaptive resources enabled

::::{tab-set}

> :::{tab-item} {{stack}}

Reviewer: prefer not using cloud, but you can if you really want

```suggestion
:::{tab-item} ECH, ECK, ECE
```

```suggestion
:::{tab-item} Cloud
```

| Level | Allocations | Threads | vCPUs |
| --- | --- | --- | --- |
| Low | 0 to 2 if available, dynamically | 1 | 0 to 2 if available, dynamically |
| Medium | 1 to 32 dynamically | 1 | 1 to the smaller of 32 or the limit set in the Cloud console, dynamically |

> | High | 1 to limit set in the Cloud console *, dynamically | 1 | 1 to limit set in the Cloud console, dynamically |

Reviewer: ECK does not have a cloud console ... does this still apply?

Author: Thanks for pointing this out! @eedugon helped me figure out how ECK handles adaptive resources. Based on that, I updated the tables to remove ECK and instead clarified its behavior in the sentence below:

	“If you use a self-managed cluster or ECK, vCPU level ranges are derived from the total_ml_processors and max_single_ml_node_processors values.”

\* The Cloud console doesn’t directly set an allocations limit; it only sets a vCPU limit. This vCPU limit indirectly determines the number of allocations, calculated as the vCPU limit divided by the number of threads.

:::

:::{tab-item} {{serverless-short}}

| Level | Allocations | Threads | VCUs |
| --- | --- | --- | --- |
| Low | 0 to 2 dynamically | 1 | 0 to 16 dynamically |
| Medium | 1 to 32 dynamically | 1 | 8 to 256 dynamically |
| High | 1 to 512 for Search<br> 1 to 128 for Security and Observability<br> | 1 | 8 to 4096 for Search<br> 8 to 1024 for Security and Observability<br> |

:::

::::

#### Adaptive resources disabled

::::{tab-set}

> :::{tab-item} {{stack}}

Reviewer: ```suggestion
:::{tab-item} ECH, ECK, ECE
```

```suggestion
:::{tab-item} Cloud
```

| Level | Allocations | Threads | vCPUs |
| --- | --- | --- | --- |
| Low | 2 if available, otherwise 1, statically | 1 | 2 if available |

> | Medium | the smaller of 32 or the limit set in the Cloud console, statically | 1 | 32 if available |

Reviewer: same note about the cloud console

applies_to:
  deployment: 
    ess:
    eck:
    ece:
  serverless: 
---

# Trained model autoscaling

You can enable autoscaling for each of your trained model deployments. Autoscaling allows {{es}} to automatically adjust the resources the model deployment can use based on the workload demand.

There are two ways to enable autoscaling:

* through APIs by enabling adaptive allocations
* in {{kib}} by enabling adaptive resources

::::{important}
To fully leverage model autoscaling in {{ech}}, {{ece}}, and {{eck}}, it is highly recommended to enable [{{es}} deployment autoscaling](../../deploy-manage/autoscaling.md).
::::

Trained model autoscaling is available for {{serverless-short}}, {{ech}}, {{ece}}, and {{eck}} deployments. In serverless deployments, processing power is managed differently across Search, Observability, and Security projects, which impacts their costs and resource limits.

:::{admonition} Trained model auto-scaling for self-managed deployments
The available resources of self-managed deployments are static, so trained model autoscaling is not applicable. However, available resources are still segmented based on the settings described in this section.

> :::

Reviewer: consider putting this note on the overview page too

:::{image} ../../images/machine-learning-ml-nlp-deployment-id-elser-v2.png
:alt: ELSER deployment with adaptive resources enabled.
:class: screenshot
:width: 500px
:::

In {{serverless-full}}, Search projects are given access to more processing resources, while Security and Observability projects have lower limits. This difference is reflected in the UI configuration: Search projects have higher resource limits compared to Security and Observability projects to accommodate their more complex operations.

On {{serverless-short}}, adaptive allocations are automatically enabled for all project types. However, the "Adaptive resources" control is not displayed in Kibana for Observability and Security projects.

## Model deployment resource matrix [model-deployment-resource-matrix]

The used resources for trained model deployments depend on three factors:

* your cluster environment ({{serverless-short}}, Cloud (ECE, ECK, ECH), or self-managed)
* the use case you optimize the model deployment for (ingest or search)
* whether model autoscaling is enabled with adaptive allocations/resources to have dynamic resources, or disabled for static resources

> If you use a self-managed cluster or ECK, vCPUs level ranges are derived from the `total_ml_processors` and `max_single_ml_node_processors` values. Use the [get {{ml}} info API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-ml-info) to check these values. The following tables show you the number of allocations, threads, and vCPUs available in ECE, ECK, and ECH when adaptive resources are enabled or disabled.

Reviewer: maybe we take out eck here

```suggestion
If you use a self-managed cluster or ECK, vCPUs level ranges are derived from the `total_ml_processors` and `max_single_ml_node_processors` values. Use the [get {{ml}} info API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-ml-info) to check these values. 

The following tables show you the number of allocations, threads, and vCPUs available in ECE and ECH when adaptive resources are enabled or disabled.
```

mapped_urls:
  - https://www.elastic.co/guide/en/serverless/current/general-ml-nlp-auto-scale.html
  - https://www.elastic.co/guide/en/serverless/current/general-ml-nlp-auto-scale.html
applies_to:
  stack: ga
  serverless: ga
---

# Trained model autoscaling

You can enable autoscaling for each of your trained model deployments. Autoscaling allows {{es}} to automatically adjust the resources the model deployment can use based on the workload demand.

There are two ways to enable autoscaling:

* through APIs by enabling adaptive allocations
* in {{kib}} by enabling adaptive resources

> ::::{important}
> To fully leverage model autoscaling, it is highly recommended to enable [{{es}} deployment autoscaling](../../deploy-manage/autoscaling.md).
> ::::

Reviewer: this is not needed for serverless

```suggestion
::::{important}
To fully leverage model autoscaling in {{ech}}, {{ece}}, and {{eck}}, it is highly recommended to enable [{{es}} deployment autoscaling](../../deploy-manage/autoscaling.md).
::::
```

> Trained model autoscaling is available for both serverless and Cloud deployments. In serverless deployments, processing power is managed differently across Search, Observability, and Security projects, which impacts their costs and resource limits.

Reviewer: @szabosteve later on this page, we say that the used resources depend on your cluster env (serverless, cloud, onprem)

<img width="640" alt="image" src="https://github.com/user-attachments/assets/9cb29746-51a4-4fbd-aa47-4bcdbe580304" />

do you know if this means TM autoscaling is available for self-managed, even though regular autoscaling is not?

Reviewer 2: @shainaraskas Trained model autoscaling is unavailable for on-prem, but the vCPU level ranges can still be interpreted in the context. This is why on-prem is on the list of applicable environments for model deployment resources, as the docs point out a bit later on the page:

```
If you use Elasticsearch on-premises, vCPUs level ranges are derived from the `total_ml_processors` 
and `max_single_ml_node_processors` values. Use the get machine learning info API 
to check these values.
```

The used resources for trained model deployment are static for on-prem. The exact numbers for `low`, `medium`, `high` depend on the above-mentioned settings. In short, the system cannot add new resources to your cluster based on the model demand if you are on-prem, but your available resources are still segmented (low, medium, high) based on your ML settings.

$$$nlp-model-adaptive-resources$$$

You can enable adaptive resources for your models when starting or updating the model deployment. Adaptive resources make it possible for {{es}} to scale up or down the available resources based on the load on the process. This can help you to manage performance and cost more easily. When adaptive resources are enabled, the number of vCPUs that the model deployment uses is set automatically based on the current load. When the load is high, the number of vCPUs that the process can use is automatically increased. When the load is low, the number of vCPUs that the process can use is automatically decreased.

You can choose from three levels of resource usage for your trained model deployment; autoscaling will occur within the selected level’s range.

Refer to the tables in the [Model deployment resource matrix](#model-deployment-resource-matrix) section to find out the setings for the level you selected.

:::{image} ../../images/machine-learning-ml-nlp-deployment-id-elser-v2.png
:alt: ELSER deployment with adaptive resources enabled.
:class: screenshot

> :::

Reviewer: this picture is really big and should be [sized down](https://elastic.github.io/docs-builder/syntax/images/#block-level-images)

Search projects are given access to more processing resources, while Security and Observability projects have lower limits. This difference is reflected in the UI configuration: Search projects have higher resource limits compared to Security and Observability projects to accommodate their more complex operations.

On Serverless, adaptive allocations are automatically enabled for all project types. However, the "Adaptive resources" control is not displayed in Kibana for Observability and Security projects.

## Model deployment resource matrix [model-deployment-resource-matrix]

The used resources for trained model deployments depend on three factors:

> * your cluster environment (Serverless, Cloud, or on-premises)

Reviewer: after we determine what this means, we should clarify

Author: Based on Istvan’s answer, I have added a clarifying sentence to the document explaining how on-premises is relevant to this page:
"The available resources of on-premises deployments are static, trained model autoscaling is not applicable. However, available resources are still segmented based on the settings described in this section."

* the use case you optimize the model deployment for (ingest or search)
* whether model autoscaling is enabled with adaptive allocations/resources to have dynamic resources, or disabled for static resources

If you use {{es}} on-premises, vCPUs level ranges are derived from the `total_ml_processors` and `max_single_ml_node_processors` values. Use the [get {{ml}} info API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-ml-info) to check these values. The following tables show you the number of allocations, threads, and vCPUs available in Cloud when adaptive resources are enabled or disabled.

::::{note}
On Serverless, adaptive allocations are automatically enabled for all project types. However, the "Adaptive resources" control is not displayed in {{kib}} for Observability and Security projects.
::::

> ### Deployments in Cloud optimized for ingest [_deployments_in_cloud_optimized_for_ingest]

Reviewer: I suggest merging the cloud and serverless sections and using tabs (e.g. a tab set under each H4, in two H3 sections - one of ingest and one for search). it will be easier to find what people are looking for. 

if these apply to elastic cloud hosted, and not some other meaning of cloud,  then you should adjust these titles. you should adjust them either way to make the scope clear.

> ```{applies_to}
> stack: {{ecloud}}
> ```

Reviewer: ````suggestion
```{applies_to}
ech: all
```
````

In case of ingest-optimized deployments, we maximize the number of model allocations.

#### Adaptive resources enabled [_adaptive_resources_enabled]

| Level | Allocations | Threads | vCPUs |
| --- | --- | --- | --- |
| Low | 0 to 2 if available, dynamically | 1 | 0 to 2 if available, dynamically |
| Medium | 1 to 32 dynamically | 1 | 1 to the smaller of 32 or the limit set in the Cloud console, dynamically |
| High | 1 to limit set in the Cloud console *, dynamically | 1 | 1 to limit set in the Cloud console, dynamically |

> * The Cloud console doesn’t directly set an allocations limit; it only sets a vCPU limit. This vCPU limit indirectly determines the number of allocations, calculated as the vCPU limit divided by the number of threads.

Reviewer: this asterisk is showing up as a bullet. 

```suggestion
\* The Cloud console doesn’t directly set an allocations limit; it only sets a vCPU limit. This vCPU limit indirectly determines the number of allocations, calculated as the vCPU limit divided by the number of threads.
```

#### Adaptive resources disabled [_adaptive_resources_disabled]

| Level | Allocations | Threads | vCPUs |
| --- | --- | --- | --- |
| Low | 2 if available, otherwise 1, statically | 1 | 2 if available |
| Medium | the smaller of 32 or the limit set in the Cloud console, statically | 1 | 32 if available |
| High | Maximum available set in the  Cloud console *, statically | 1 | Maximum available set in the Cloud console, statically |

> * The Cloud console doesn’t directly set an allocations limit; it only sets a vCPU limit. This vCPU limit indirectly determines the number of allocations, calculated as the vCPU limit divided by the number of threads.

Reviewer: not sure this is the right way to escape an asterisk - should check it renders properly

```suggestion
\* The Cloud console doesn’t directly set an allocations limit; it only sets a vCPU limit. This vCPU limit indirectly determines the number of allocations, calculated as the vCPU limit divided by the number of threads.
```

### Deployments in Cloud optimized for search [_deployments_in_cloud_optimized_for_search]

> ```{applies_to}
> stack: {{ecloud}}
> ```

Reviewer: ````suggestion
```{applies_to}
ech: all
```
````

In case of search-optimized deployments, we maximize the number of threads. The maximum number of threads that can be claimed depends on the hardware your architecture has.

#### Adaptive resources enabled [_adaptive_resources_enabled_2]

| Level | Allocations | Threads | vCPUs |
| --- | --- | --- | --- |
| Low | 1 | 2 | 2 |
| Medium | 1 to 2 (if threads=16) dynamically | maximum that the hardware allows (for example, 16) | 1 to 32 dynamically |
| High | 1 to limit set in the Cloud console *, dynamically | maximum that the hardware allows (for example, 16) | 1 to limit set in the Cloud console, dynamically |

> * The Cloud console doesn’t directly set an allocations limit; it only sets a vCPU limit. This vCPU limit indirectly determines the number of allocations, calculated as the vCPU limit divided by the number of threads.

Reviewer: ```suggestion
\* The Cloud console doesn’t directly set an allocations limit; it only sets a vCPU limit. This vCPU limit indirectly determines the number of allocations, calculated as the vCPU limit divided by the number of threads.
```

#### Adaptive resources disabled [_adaptive_resources_disabled_2]

| Level | Allocations | Threads | vCPUs |
| --- | --- | --- | --- |
| Low | 1 if available, statically | 2 | 2 if available |
| Medium | 2 (if threads=16) statically | maximum that the hardware allows (for example, 16) | 32 if available |
| High | Maximum available set in the Cloud console *, statically | maximum that the hardware allows (for example, 16) | Maximum available set in the Cloud console, statically |

\* The Cloud console doesn’t directly set an allocations limit; it only sets a vCPU limit. This vCPU limit indirectly determines the number of allocations, calculated as the vCPU limit divided by the number of threads.

### Deployments on serverless optimized for ingest [deployments-on-serverless-optimized-for-ingest]

> ```{applies_to}
> stack: {{serverless-short}}
> ```

Reviewer: ````suggestion
```{applies_to}
serverless: all
```
````

In case of ingest-optimized deployments, we maximize the number of model allocations.

#### Adaptive resources enabled [adaptive-resources-enabled]

| Level | Allocations | Threads | VCUs |
| --- | --- | --- | --- |
| Low | 0 to 2 dynamically | 1 | 0 to 16 dynamically |
| Medium | 1 to 32 dynamically | 1 | 8 to 256 dynamically |
| High | 1 to 512 for Search<br> 1 to 128 for Security and Observability<br> | 1 | 8 to 4096 for Search<br> 8 to 1024 for Security and Observability<br> |

#### Adaptive resources disabled (Search only) [adaptive-resources-disabled-search-only]

| Level | Allocations | Threads | VCUs |
| --- | --- | --- | --- |
| Low | Exactly 2 | 1 | 16 |
| Medium | Exactly 32 | 1 | 256 |
| High | 512 for Search<br> No static allocations for Security and Observability<br> | 1 | 4096 for Search<br> No static allocations for Security and Observability<br> |

### Deployments on serverless optimized for Search [deployments-on-serverless-optimized-for-search]

> ```{applies_to}
> stack: {{serverless-short}}
> ```

Reviewer: ````suggestion
```{applies_to}
serverless: all
```
````
